feat: show bigquery jobId when executing the job

ashish10alex opened the pull request:
Solves: https://github.com/dataform-co/dataform/issues/1944

Shows BigQuery jobId against each ran model as follows. Happy to change the format the jobId is displayed 

<img width="1215" alt="image" src="https://github.com/user-attachments/assets/0ecf0c7c-4174-49da-b88d-dec068bd2742" />


**Tests**

- [x] `bazel test //core/...` passes 
- [x] JobId shows against the correct model 
- [x] JobId shows against failed model / assertion run 
- [x] Linting by running `./scripts/lint`


**P.S.**

It will be good to have have the jobIds as hyperlinks but most default terminal emulators do not support this. From the terminals I have only Kitty supported it. Default MacOS terminal and Ghostty terminal did not support it.

Ceridan commented on cli/api/commands/run.ts:
```diff
@@ -441,6 +441,11 @@ export class Runner {
           ? dataform.TaskResult.ExecutionStatus.CANCELLED
           : dataform.TaskResult.ExecutionStatus.FAILED;
         taskResult.errorMessage = `${this.graph.projectConfig.warehouse} error: ${e.message}`;
+        taskResult.metadata = {
```
Could we please set `metadata` it only when `e.metadata?.bigquery?.jobId` is defined? At the moment metadata is optional and it makes sense to set it only if we actually have some value.

It will also make integration tests pipeline green.

ashish10alex replied:
thanks for the feedback  @Ceridan . I have made the requested change.

Ceridan reviewed (changes requested):
@ashish10alex, thanks a lot for the contribution!

I requested a minor change and with it in-place will happy to accept the PR.

Ceridan commented on cli/api/dbadapters/bigquery.ts:
```diff
@@ -356,12 +356,26 @@ export class BigQueryDbAdapter implements IDbAdapter {
+                try {
+                  const [jobMetadata] = await job[0].getMetadata();
+                  if (jobMetadata && jobMetadata.jobReference && jobMetadata.jobReference.jobId) {
+                    e.metadata = {
```
Adding a `metadata` field is a little bit unexpected here and can be confusing. Because we use TypeScript it makes sense to define a new interface in `/cli/api/dbadapters/index.ts` like:

```typescript
interface BigQueryError extends Error {
  metadata?: dataform.IExecutionMetadata
}
```
and construct it here?

The same for the check below.

ashish10alex replied:
added and exported `BigQueryError` as suggested

Ceridan commented on cli/api/dbadapters/bigquery.ts:
```diff
@@ -374,7 +388,15 @@ export class BigQueryDbAdapter implements IDbAdapter {
                   const [jobMetadata] = await job[0].getMetadata();
                   if (!!jobMetadata.status?.errorResult) {
-                    reject(new Error(jobMetadata.status.errorResult.message));
+                    const error: any = new Error(jobMetadata.status.errorResult.message);
```
Because we need to construct the error twice, I suggest to extract a helper method which accepts `jobMetadata` as argument, and returns `BigQueryError`.

ashish10alex replied:
created a new function `createBigQueryError`

Ceridan reviewed (changes requested):
Thank you for the quick response and fixes. I think we can make it a slightly better, could you please take a look at my comments on error construction?

ashish10alex commented:
> Thank you for the quick response and fixes. I think we can make it a slightly better, could you please take a look at my comments on error construction?

thanks @Ceridan  for the suggestion to improve the code. I have made the requested changes now

Ceridan reviewed (changes requested):
The code looks great, thanks a lot!

The one last blocker: could you please fix the linter issues?

```
ERROR: /usr/local/google/home/mkumachev/repos/dataform-co/dataform/cli/api/dbadapters/bigquery.ts:6:51 - Named imports must be alphabetized.
ERROR: /usr/local/google/home/mkumachev/repos/dataform-co/dataform/cli/api/dbadapters/index.ts:11:18 - interface name must start with a capitalized I
```

You may reproduce them by running:

```
./scripts/lint
```

I know that at the moment our toolchain is not very contributor-friendly but I have some plans to improve it in the near future.

ashish10alex commented:
>  could you please fix the linter issues?

Thanks @Ceridan , I appreciate you guiding me through  the process :) 

I have fixed the lining error and also added the `./scripts/lint` to the `contributing.md` file. 

**Tests**

- [x] Linting

![CleanShot 2025-03-31 at 09 42 27@2x](https://github.com/user-attachments/assets/606ece5e-501f-44f4-bd2a-61e159ea2ac8)

Ceridan reviewed (approved):
Thank you!

I will merge the changes, and plan to release the new CLI version this week which will include the changes.
